### PR TITLE
describe module inheritance

### PR DESCRIPTION
**Pull request by mikz**
closes https://github.com/3scale/apicast/issues/308

* improve documentation in the example
* provide whole module (blacklist.lua)
* test that example module

and polish a bit balancer interface, so we can easily modify it from modules

### CODE REVIEW

File: examples/custom-module/README.md [hunk 1]
```diff
@@ -26,3 +26,39 @@ And see in the apicast output:
...
+There is example module of IP blacklist in [`blacklist.lua`](blacklist.lua).
+
+To honour the module inheritance, but still be able to override some methods from the `apicast` module, you'll
+need some clever use of metatables. Here a recommended skeleton of the module inheritance:
+
+```lua
+-- load and initialize the parent module
+local apicast = require('apicast').new()
+
+-- this is used in User-Agent
```
Comment (davidor) [inline examples/custom-module/README.md]: I'm missing something. I don't see how user-agent is related to this module.

File: examples/custom-module/README.md [hunk 2]
```diff
@@ -26,3 +26,39 @@ And see in the apicast output:
 ```
 2016/11/16 16:52:00 [warn] 98009#0: *5 [lua] verbose.lua:7: call(): upstream response time: 0.001 upstream connect time: 0.000 while logging request, client: 127.0.0.1, server: _, request: "GET /?user_key=foo HTTP/1.1", upstream: "http://127.0.0.1:8081/?user_key=foo", host: "echo"
 ```
+
+## Writing own module
+
+There is example module of IP blacklist in [`blacklist.lua`](blacklist.lua).
+
+To honour the module inheritance, but still be able to override some methods from the `apicast` module, you'll
+need some clever use of metatables. Here a recommended skeleton of the module inheritance:
```
Comment (davidor) [inline examples/custom-module/README.md]: typo: here's a recommended...

File: examples/custom-module/README.md [hunk 3]
```diff
@@ -26,3 +26,39 @@ And see in the apicast output:
...
+-- so in this case the inheritance works like mod -> _M -> apicast
+local mt = { __index = setmetatable(_M, { __index = apicast }) }
+
+function _M.new()
+  -- this method is going to get called after this file is required
+  -- so create new table for internal state (global) and set the metatable for inheritance
+  return setmetatable({}, mt)
+end
+
+function _M.log()
```
Comment (davidor) [inline examples/custom-module/README.md]: To me, this seems to indicate that defining `log()` is required, which is not the case. I would specify somewhere in this file that the purpose of this module is to log some things that apicast by itself does not, and that's why we need to override `log()`.

File: examples/custom-module/README.md (hunk 1 above)
Comment (mikz) [inline examples/custom-module/README.md]: Well the used module _NAME field is going to get used in User-Agent header from APIcast connecting to backend.

File: examples/custom-module/README.md (hunk 3 above)
Comment (mikz) [inline examples/custom-module/README.md]: Would a comment be enough?

Comment (davidor) [inline examples/custom-module/README.md]: yes

File: examples/custom-module/README.md [hunk 4]
```diff
@@ -26,3 +26,40 @@ And see in the apicast output:
...
+```lua
+-- load and initialize the parent module
+local apicast = require('apicast').new()
+
+-- _NAME and _VERSION are used in User-Agent when connecting to the Service Management API
+local _M = { _VERSION = '0.0', _NAME = 'Example Module' }
+-- define a table, that is going to be this module metatable
+-- if your table does not define a property, __index is going to get used
+-- and so on until there are no metatables to check
+-- so in this case the inheritance works like mod -> _M -> apicast
```
Comment (mayorova) [inline examples/custom-module/README.md]: what is `mod`?

**Comment by mayorova** [comment]
Nice explanation! Just one thing to clarify, in a comment above...

File: examples/custom-module/README.md (hunk 4 above)
Comment (mikz) [inline examples/custom-module/README.md]: That would be instance of the module. Not really sure how to write it.

Comment (mikz) [inline examples/custom-module/README.md]: Would something like `local instance = _M.new() -> _M -> apicast` work for you?

Comment (mayorova) [inline examples/custom-module/README.md]: OK, much clearer now :)

How about:
>so in this case the inheritance works like this: local instance created with _M.new() -> _M -> apicast

Comment (mikz) [inline examples/custom-module/README.md]: Perfect. Fixed.

**Pull request merged**